src/Microsoft.AspNet.OData.Shared/EdmEntityObject.cs

```diff
@@ -42,6 +46,62 @@ public EdmEntityObject(IEdmEntityTypeReference edmType)
         public EdmEntityObject(IEdmEntityType edmType, bool isNullable)
             : base(edmType, isNullable)
         {
+            PersistentInstanceAnnotationsContainer = new ODataInstanceAnnotationContainer();
+            TransientInstanceAnnotationContainer = new ODataInstanceAnnotationContainer();
+        }
+
+        /// <summary>
+        /// Gets or sets the instance annotation container to hold transient annotations.
+        /// </summary>
+        internal IODataInstanceAnnotationContainer TransientInstanceAnnotationContainer { get; set; }
+
+        /// <summary>
+        /// Gets or sets the instance annotation container to hold persistent annotations.
+        /// </summary>
+        public IODataInstanceAnnotationContainer PersistentInstanceAnnotationsContainer { get; set; }
+
+        /// <summary>
+        /// Gets or sets the container to hold ODataId.
+        /// </summary>
+        public IODataIdContainer ODataIdContainer { get; set; }
+
+        /// <summary>
+        /// Gets or sets the OData path for the item.
+        /// </summary>
+        internal ODataPath ODataPath { get; set; }
+
+        /// <summary>
+        /// Gets the kind of object within the delta payload.
+        /// </summary>
+        public virtual EdmDeltaEntityKind DeltaKind { get { return EdmDeltaEntityKind.Entry; } }
+
+        /// <summary>
+        /// Adds the specified data modification exception.
+        /// </summary>
+        public void AddDataException(DataModificationExceptionType dataModificationException)
+        {
+            if (TransientInstanceAnnotationContainer == null)
+            {
+                throw Error.ArgumentNull(nameof(TransientInstanceAnnotationContainer));
+            }
```

Comment:
since you initialize in the constructor, why do we need to check  the null again here?